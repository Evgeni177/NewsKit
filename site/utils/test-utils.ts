import {
  renderToFragmentWithThemeFactory,
  renderWithThemeFactory,
} from 'newskit/test/test-utils';
import {docsThemeLight} from '../theme/doc-theme';

export const renderWithTheme = renderWithThemeFactory(docsThemeLight);

export const renderToFragmentWithTheme = renderToFragmentWithThemeFactory(
  docsThemeLight,
);
