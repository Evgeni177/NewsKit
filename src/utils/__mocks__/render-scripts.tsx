import React from 'react';
import {RenderScriptsProps} from '../render-scripts';

export const RenderScripts: React.FC<RenderScriptsProps> = ({
  scripts,
  reactHelmet,
}) => (
  <div>
    RenderScripts Component with props:
    {JSON.stringify({scripts, reactHelmet: typeof reactHelmet}, null, 2)}
  </div>
);
