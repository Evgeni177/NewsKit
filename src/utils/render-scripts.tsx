import React from 'react';
import {getSSRId} from './get-ssr-id';

interface ExternalScriptData {
  src: string;
}

interface InlineScriptData {
  content: string;
}

type ScriptData = ExternalScriptData | InlineScriptData;
interface HelmetScriptProps {
  type: 'text/javascript';
  src?: string;
  innerHTML?: string;
}

const isExternalScriptData = (
  scriptData: ScriptData,
): scriptData is ExternalScriptData => 'src' in scriptData;

export interface RenderScriptsReactHelmetProp {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  reactHelmet?: React.ComponentType<{script?: Array<any>}>;
}

export interface RenderScriptsProps extends RenderScriptsReactHelmetProp {
  scripts: ScriptData[];
}

export const RenderScripts: React.FC<RenderScriptsProps> = ({
  scripts,
  reactHelmet: ReactHelmet,
}) => {
  if (ReactHelmet) {
    return (
      <ReactHelmet
        script={[
          ...scripts.map(scriptData => {
            const scriptProps = {
              type: 'text/javascript',
            } as HelmetScriptProps;

            if (isExternalScriptData(scriptData)) {
              scriptProps.src = scriptData.src;
            } else {
              scriptProps.innerHTML = scriptData.content;
            }

            return scriptProps;
          }),
        ]}
      />
    );
  }

  return (
    <React.Fragment>
      {scripts.map(scriptData => {
        const scriptProps = isExternalScriptData(scriptData)
          ? {
              key: scriptData.src,
              src: scriptData.src,
            }
          : {
              key: getSSRId(),
              dangerouslySetInnerHTML: {
                __html: scriptData.content,
              },
            };
        return <script type="text/javascript" {...scriptProps} />;
      })}
    </React.Fragment>
  );
};
