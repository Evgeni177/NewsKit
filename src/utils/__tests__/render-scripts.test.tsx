import React from 'react';
import {renderToFragment} from '../../test/test-utils';
import {RenderScripts, RenderScriptsProps} from '../render-scripts';

describe('RenderScripts', () => {
  const scripts: RenderScriptsProps['scripts'] = [
    {
      src: 'https://example.com/my-javascript-file.js',
      content: '',
    },
    {
      content: `(function () { alert('some injected script') })();`,
    },
    {
      src: 'https://example.com/my-javascript-file2.js',
      content: '',
    },
    {
      content: `(function () { alert('some more injected script') })();`,
    },
  ];

  test('renders inline and external scripts as script tags', () => {
    const fragment = renderToFragment(<RenderScripts scripts={scripts} />);
    expect(fragment).toMatchSnapshot();
  });

  test('renders inline and external scripts via ReactHelmet', () => {
    const MockReactHelmet: React.FC = props => (
      <div>MockReactHelmet props: {JSON.stringify(props, null, 2)}</div>
    );

    const fragment = renderToFragment(
      <RenderScripts scripts={scripts} reactHelmet={MockReactHelmet} />,
    );
    expect(fragment).toMatchSnapshot();
  });
});
