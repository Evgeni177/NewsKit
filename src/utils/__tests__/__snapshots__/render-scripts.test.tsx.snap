// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RenderScripts renders inline and external scripts as script tags 1`] = `
<DocumentFragment>
  <script
    src="https://example.com/my-javascript-file.js"
    type="text/javascript"
  />
  <script
    type="text/javascript"
  >
    (function () { alert('some injected script') })();
  </script>
  <script
    src="https://example.com/my-javascript-file2.js"
    type="text/javascript"
  />
  <script
    type="text/javascript"
  >
    (function () { alert('some more injected script') })();
  </script>
</DocumentFragment>
`;

exports[`RenderScripts renders inline and external scripts via ReactHelmet 1`] = `
<DocumentFragment>
  <div>
    MockReactHelmet props: {
  "script": [
    {
      "type": "text/javascript",
      "src": "https://example.com/my-javascript-file.js"
    },
    {
      "type": "text/javascript",
      "innerHTML": "(function () { alert('some injected script') })();"
    },
    {
      "type": "text/javascript",
      "src": "https://example.com/my-javascript-file2.js"
    },
    {
      "type": "text/javascript",
      "innerHTML": "(function () { alert('some more injected script') })();"
    }
  ]
}
  </div>
</DocumentFragment>
`;
